snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java

```diff
@@ -96,83 +91,81 @@ private boolean applyInternal(final SnapshotChunk snapshotChunk) throws IOExcept
-      return true;
+      throw new SnapshotWriteException(
+          String.format(
+              "Expected to have checksum %d for snapshot chunk %s (%s), but calculated %d",
+              expectedChecksum, chunkName, snapshotId, actualChecksum));
     }
-    return false;
   }
 
-  private boolean isSnapshotChecksumInvalid(final long currentSnapshotChecksum) {
+  private void checkSnapshotChecksumIsValid(final long currentSnapshotChecksum)
+      throws SnapshotWriteException {
     if (expectedSnapshotChecksum == Long.MIN_VALUE) {
       expectedSnapshotChecksum = currentSnapshotChecksum;
     }
 
     if (expectedSnapshotChecksum != currentSnapshotChecksum) {
-      LOGGER.warn(
-          "Expected snapshot chunk with equal snapshot checksum {}, but got chunk with snapshot checksum {}.",
-          expectedSnapshotChecksum,
-          currentSnapshotChecksum);
-      return true;
+      throw new SnapshotWriteException(
+          String.format(
+              "Expected snapshot chunk with equal snapshot checksum %d, but got chunk with snapshot checksum %d.",
+              expectedSnapshotChecksum, currentSnapshotChecksum));
     }
-    return false;
   }
 
-  private boolean isTotalCountInvalid(final int currentTotalCount) {
+  private void checkTotalCountIsValid(final int currentTotalCount) throws SnapshotWriteException {
     if (expectedTotalCount == Integer.MIN_VALUE) {
       expectedTotalCount = currentTotalCount;
     }
 
     if (expectedTotalCount != currentTotalCount) {
-      LOGGER.warn(
-          "Expected snapshot chunk with equal snapshot total count {}, but got chunk with total count {}.",
-          expectedTotalCount,
-          currentTotalCount);
-      return true;
+      throw new SnapshotWriteException(
+          String.format(
+              "Expected snapshot chunk with equal snapshot total count %d, but got chunk with total count %d.",
+              expectedTotalCount, currentTotalCount));
     }
-    return false;
   }
 
-  private boolean isSnapshotIdInvalid(final String snapshotId) {
+  private void checkSnapshotIdIsValid(final String snapshotId) throws SnapshotWriteException {
     final var receivedSnapshotId = FileBasedSnapshotMetadata.ofFileName(snapshotId);
-    if (receivedSnapshotId.isEmpty()) {
-      return true;
+    if (receivedSnapshotId.isEmpty() || metadata.compareTo(receivedSnapshotId.get()) != 0) {
+      throw new SnapshotWriteException(
+          String.format("Snapshot file name '%s' has unexpected format", snapshotId));
     }
-    return metadata.compareTo(receivedSnapshotId.get()) != 0;
   }
 
-  private boolean writeReceivedSnapshotChunk(
-      final SnapshotChunk snapshotChunk, final Path snapshotFile) {
+  private void writeReceivedSnapshotChunk(
+      final SnapshotChunk snapshotChunk, final Path snapshotFile) throws SnapshotWriteException {
     try (var channel =
         FileChannel.open(snapshotFile, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)) {
```

Comment:
Out of scope, but since we're here and you wrote this: would it make sense to use `CREATE, WRITE, TRUNCATE` instead? `CREATE_NEW` will cause this to fail if the file already exists. Would we care if we were overwriting it? Probably not? Anyhow, it probably doesn't happen, but it was just a thought.

Reply:
I'm not sure if that situation can actually occur but I think anyway it would be an easy change so we could do that